ddtrace/appsec/ddwaf/__init__.py

```diff
@@ -109,19 +110,24 @@ def run(
 
             ctx = ddwaf_context_init(self._handle)
             if ctx == 0:
-                raise RuntimeError
-            try:
-                wrapper = ddwaf_object(data)
-                error, result = py_ddwaf_run(ctx, wrapper, timeout_ms * 1000)
-                return DDWaf_result(
-                    result.data.decode("UTF-8") if result.data else None,
-                    [result.actions.array[i].decode("UTF-8") for i in range(result.actions.size)],
-                    result.total_runtime / 1e3,
-                    (time.time() - start) * 1e6,
-                )
-            finally:
-                ddwaf_result_free(ctypes.byref(result))
-                ddwaf_context_destroy(ctx)
+                LOGGER.error("DDWaf failure to create the context")
+                return DDWaf_result(None, [], 0, (time.time() - start) * 1e6)
+
+            result = ddwaf_result()
+            wrapper = ddwaf_object(data)
+            error = ddwaf_run(ctx, wrapper, ctypes.byref(result), timeout_ms * 1000)
+            if error:
+                LOGGER.warning("DDWAF error: %d", error)
+
+            libddwaf_result = DDWaf_result(
+                result.data.decode("UTF-8", errors="ignore") if hasattr(result, "data") and result.data else None,
+                [result.actions.array[i].decode("UTF-8", errors="ignore") for i in range(result.actions.size)],
+                result.total_runtime / 1e3,
+                (time.time() - start) * 1e6,
+            )
```

Comment:
everything in `result` gets free'd right? so the only reason this does not cause a bad pointer/reference is because everything is transformed/copied right?

if you didn't call `.decode()` on `result.data`, then we'd have issues?


if so, maybe a comment or helper would be better, I could see an easy mistake to end up with an invalid pointer to something here.

if not, feel free to resolve with no action!

Reply:
I wonder if we should still have `try/finally` here, since if we end up with an error in any of the transformations when we could end up leaking memory.

The only issues I potentially see are data that aren't actually bytes, or is `size` is wrong.

```python
try:
    return DDWaf_result(...)
finally:
    ddwaf_result_free(ctypes.byref(result))
    ddwaf_context_destroy(ctx)
```


In fact, `ctx` could leak as well, if there is an error anywhere between L116 and the end of the function.

maybe better as:


```python
ctx = ddwaf_context_init(self._handle)
try:
    result = ddwaf_result()
    try:
        return DDWaf_result(...)
    finally:
        ddwaf_result_free(ctypes.byref(result))
finally:
    ddwaf_context_destroy(ctx)
```

I know it isn't very pretty, but we really want to be sure we don't end up with a memory leak.